Library/Homebrew/utils/spdx.rb

```diff
@@ -5,19 +5,162 @@
 module SPDX
   module_function
 
-  JSON_PATH = (HOMEBREW_LIBRARY_PATH/"data/spdx.json").freeze
+  LICENSES_JSON_PATH = (HOMEBREW_DATA_PATH/"spdx_licenses.json").freeze
+  EXCEPTIONS_JSON_PATH = (HOMEBREW_DATA_PATH/"spdx_exceptions.json").freeze
   API_URL = "https://api.github.com/repos/spdx/license-list-data/releases/latest"
 
-  def spdx_data
-    @spdx_data ||= JSON.parse(JSON_PATH.read)
+  def license_data
+    @license_data ||= JSON.parse LICENSES_JSON_PATH.read
+  end
+
+  def exception_data
+    @exception_data ||= JSON.parse EXCEPTIONS_JSON_PATH.read
   end
 
   def latest_tag
     @latest_tag ||= GitHub.open_api(API_URL)["tag_name"]
   end
 
-  def download_latest_license_data!(to: JSON_PATH)
-    data_url = "https://raw.githubusercontent.com/spdx/license-list-data/#{latest_tag}/json/licenses.json"
-    curl_download(data_url, to: to, partial: false)
+  def download_latest_license_data!(licenses_to: LICENSES_JSON_PATH, exceptions_to: EXCEPTIONS_JSON_PATH)
+    data_url = "https://raw.githubusercontent.com/spdx/license-list-data/#{latest_tag}/json/"
+    curl_download("#{data_url}licenses.json", to: licenses_to, partial: false)
+    curl_download("#{data_url}exceptions.json", to: exceptions_to, partial: false)
+  end
+
+  def parse_license_expression(license_expression)
+    licenses = []
+    exceptions = []
+
+    case license_expression
+    when String, Symbol
+      licenses.push license_expression
+    when Hash
+      license_expression.each_key do |key|
+        if [:any_of, :all_of].include? key
+          sub_license, sub_exception = parse_license_expression license_expression[key]
+          licenses += sub_license
+          exceptions += sub_exception
+        else
+          licenses.push key
+          exceptions.push license_expression[key][:with]
+        end
+      end
+    when Array
+      license_expression.each do |license|
+        sub_license, sub_exception = parse_license_expression license
+        licenses += sub_license
+        exceptions += sub_exception
+      end
+    end
+
+    [licenses, exceptions]
+  end
+
+  def valid_license?(license)
+    return true if license == :public_domain
+
+    license = license.delete_suffix "+"
+    license_data["licenses"].any? { |spdx_license| spdx_license["licenseId"] == license }
+  end
+
+  def deprecated_license?(license)
+    return false if license == :public_domain
+    return unless valid_license?(license)
```

Comment:
```suggestion
    return false unless valid_license?(license)
```

Reply:
It makes me uncomfortable to return `false` here because I feel like `deprecated_license?` being `false` implies that the license is okay (which it isn't in this case). That's why I like returning `nil` because it is treated as false but doesn't have the implication of being correct.

I won't fight too hard on this, though.

Reply:
A `?` method should always return a boolean, hence the `false`. Alternatively: raise an exception or similar which requires `valid_license?` to have always been called first.

Reply:
Alright, I'll just have it return false. No need to add extra complexity.